analyzers/src/SonarAnalyzer.Common/Rules/Utilities/TokenTypeAnalyzerBase.cs

```diff
@@ -105,13 +104,16 @@ protected internal abstract class TokenClassifierBase
                 SymbolKind.ArrayType,
                 SymbolKind.PointerType
             };
+            private readonly SemanticModel semanticModel;
 
             protected abstract SyntaxNode GetBindableParent(SyntaxToken token);
             protected abstract bool IsKeyword(SyntaxToken token);
             protected abstract bool IsIdentifier(SyntaxToken token);
             protected abstract bool IsNumericLiteral(SyntaxToken token);
             protected abstract bool IsStringLiteral(SyntaxToken token);
 
+            protected SemanticModel SemanticModel => semanticModel ?? throw new InvalidOperationException("The code snippet is not supposed to call the semantic model for classification.");
```

Comment:
> The code snippet is not supposed to call the semantic model for classification.

I think it's not ok for the semantic model to know how it will be used (classification). In the future, this might change. We should not throw here at all and let the call raise an exception where this is used.

Reply:
The class is called TokenClassifierBase, so it is used for classification.

This change was introduced due to an issue with the mock. I needed to change the logic in the TokenTypeAnalyzerTest.ClassifierTestHarness.cs:
analyzers/tests/SonarAnalyzer.UnitTest/Rules/Utilities/TokenTypeAnalyzerTest.ClassifierTestHarness.cs

We can not rely on the strict Mock of the semantic model because we now use some extension methods and these methods seem to either not call the underlying semantic model or call something the mock can not intercept (non-virtual methods).

Passing null for the semantic model in the tests is the only reliable way I could find and this exception is the replacement for the previous Exception raised by the strict Mock.

Reply:
So, if I understand right, the reason you throw the exception is just for testing. Wouldn't be better to add this behavior to a Fake class that derives from TokenTypeAnalizer and does this only in the test code?

Reply:
This seems like overkill to me and TokenClassifier is `sealed` at the moment. I was thinking about an internal constructor on the base class instead but it seemed to be to much hassle to me:
```cs
            protected TokenClassifierBase(SemanticModel semanticModel, bool skipIdentifiers)
            {
                this.semanticModel = semanticModel ?? throw new ArgumentNullException(nameof(semanticModel));
                this.skipIdentifiers = skipIdentifiers;
            }

            // internal ctor for testing
            protected TokenClassifierBase(bool skipIdentifiers)
            {
                this.semanticModel = null;
                this.skipIdentifiers = skipIdentifiers;
            }
```